pkg/apis/security/types.go

```diff
@@ -0,0 +1,307 @@
+	var out []byte
+	var err error
+	if out, err = json.Marshal(acl); err != nil {
+		return "", err
+	}
+	resp := string(out)
+	log.Tracef("Converted: %s", resp)
+	return resp, nil
+}
+
+func (roles *Roles) FromJson(acl string) error {
+	if err := json.Unmarshal([]byte(acl), roles); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (rolesmapping *RolesMapping) FromJson(acl string) error {
+	if err := json.Unmarshal([]byte(acl), rolesmapping); err != nil {
+		return err
+	}
+	return nil
+}
+
+//Rolesmapping are the mapping of username/groups to roles
+// root
+//  roleName
+//    expires:
+//    users:
+//    groups:
+type RolesMapping struct {
+	roleNames  map[string]RoleMapping
+	DocVersion int
+}
+
+func NewRolesMapping() *RolesMapping {
+	return &RolesMapping{roleNames: map[string]RoleMapping{}}
+}
+
+func (rolesMapping *RolesMapping) Iterate() map[string]Expirable {
+	entries := map[string]Expirable{}
+	for name, entry := range rolesMapping.roleNames {
+		copy := entry
+		entries[name] = &copy
+	}
+	return entries
+}
+
+func (rolesMapping *RolesMapping) Size() int {
+	return len(rolesMapping.roleNames)
+}
+func (rolesMapping *RolesMapping) Version() int {
+	return rolesMapping.DocVersion
+}
+func (roleMapping *RoleMapping) GetExpiresInMillis() int64 {
+	return roleMapping.ExpiresInMillis
+}
+
+func (rolesMapping *RolesMapping) Set(name string, rolemapping RoleMapping) {
+	(rolesMapping.roleNames)[name] = rolemapping
+}
+func (rolesMapping *RolesMapping) Remove(name string) {
+	delete(rolesMapping.roleNames, name)
+}
+
+type RoleMapping struct {
+	ExpiresInMillis int64    `yaml:"expires,omitempty" json:"expires,omitempty"`
+	Users           []string `yaml:"users,omitempty" json:"users,omitempty"`
+}
+
+func (rolesmapping *RolesMapping) Type() DocType {
+	return DocTypeRolesmapping
+}
+
+func (rolesmapping *RolesMapping) ToYaml() (string, error) {
+	return toYaml(rolesmapping.roleNames)
+}
+
+func (rolesmapping *RolesMapping) ToJson() (string, error) {
+	return ToJson(rolesmapping.roleNames)
+}
+
+func newSecurityDocumentPermissions(user *cl.UserInfo) IndexPermissions {
+	permissions := IndexPermissions{}
+	permissions[fix(kibanaIndexName(user))] = DocumentPermissions{
+		"*": Permissions{
+			"INDEX_KIBANA",
+		},
+	}
+	for _, project := range user.Projects {
+		permissions[fix(projectIndexName(project))] = DocumentPermissions{
+			"*": Permissions{
+				"INDEX_PROJECT",
+			},
+		}
+	}
+	return permissions
+}
+
+func fix(indexName string) string {
+	return strings.Replace(indexName, ".", "?", -1)
+}
+
+func projectIndexName(p cl.Project) string {
+	return fmt.Sprintf("project.%s.%s.*", p.Name, p.UUID)
+}
+
+func kibanaIndexName(user *cl.UserInfo) string {
+	return fmt.Sprintf(".kibana.%s", usernameHash(user))
+}
+
+func roleName(user *cl.UserInfo) string {
+	return fmt.Sprintf("gen_user_%s", usernameHash(user))
+}
+
+func usernameHash(user *cl.UserInfo) string {
+	return fmt.Sprintf("%x", sha1.Sum([]byte(user.Username)))
```

Comment:
Does this produce the exact same string as the method we use currently?  If not, do we care about reusing existing kibana indices?

Reply:
The logic doesn't exist now to do any Kibana seeding or migration but I believe we need to ensure this function produces the same result as the java version.  Kibana would normally migrate your index and we are going to have to do the same unless we want pushback from users.